apps/service_gather/lib/gather/event/handler.ex

```diff
@@ -6,37 +6,65 @@ defmodule Gather.Event.Handler do
   use Properties, otp_app: :service_gather
   require Logger
 
-  alias Gather.Extraction
+  alias Gather.{Extraction, ViewState}
   import Events, only: [extract_start: 0, extract_end: 0, dataset_delete: 0]
+  import Definition, only: [identifier: 1]
 
   getter(:endpoints, required: true)
 
   def handle_event(%Brook.Event{type: extract_start(), data: %Extract{} = extract}) do
     Logger.debug(fn -> "#{__MODULE__}: Received event #{extract_start()}: #{inspect(extract)}" end)
 
+    key = identifier(extract)
+
     Extraction.Supervisor.start_child(extract)
-    Extraction.Store.persist(extract)
+    ViewState.Extractions.persist(key, extract)
+    ViewState.Sources.persist(key, extract.source)
+    ViewState.Destinations.persist(key, extract.destination)
   end
 
   def handle_event(%Brook.Event{type: extract_end(), data: %Extract{} = extract}) do
-    Extraction.Store.mark_done(extract)
+    identifier(extract)
+    |> ViewState.Extractions.delete()
   end
 
   def handle_event(%Brook.Event{type: dataset_delete(), data: %Delete{} = delete}) do
     Logger.debug(fn ->
       "#{__MODULE__}: Received event #{dataset_delete()}: #{inspect(delete)}"
     end)
 
-    case Extraction.Store.get!(delete.dataset_id, delete.subset_id) do
-      nil ->
-        Logger.debug("No state to delete")
-        nil
+    key = identifier(delete)
+
+    case ViewState.Extractions.get(key) do
+      {:ok, nil} ->
+        delete_source(key)
+        delete_destination(key)
+        :discard
 
-      extract ->
+      {:ok, extract} ->
         Extraction.Supervisor.terminate_child(extract)
-        Source.delete(extract.source)
-        Destination.delete(extract.destination)
-        Extraction.Store.delete(delete.dataset_id, delete.subset_id)
+        delete_source(key)
+        delete_destination(key)
+        ViewState.Extractions.delete(key)
+
+      {:error, reason} ->
+        raise reason
     end
   end
+
+  defp delete_source(key) do
+    ViewState.Sources.get(key)
+    |> Ok.map(fn source ->
+      Source.delete(source)
+      ViewState.Sources.delete(key)
+    end)
+  end
+
+  def delete_destination(key) do
+    ViewState.Destinations.get(key)
+    |> Ok.map(fn dest ->
+      Destination.delete(dest)
+      ViewState.Destinations.delete(key)
+    end)
```

Comment:
i could see this being a common pattern; maybe it's worth moving these functions into the management library with the rest of the default implementations?

Reply:
Very possible. We'll see how it looks as I convert more services.